app/recipe/serializers.py

```diff
@@ -0,0 +1,44 @@
+from rest_framework import serializers
+
+from recipe.models import Ingredient, Recipe
+
+
+class IngredientSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Ingredient
+        fields = ["name"]
+        read_only_fields = ["id"]
+
+
+class RecipeSerializer(serializers.ModelSerializer):
+    ingredients = IngredientSerializer(many=True, required=False)
+
+    class Meta:
+        model = Recipe
+        fields = ["id", "name", "description", "ingredients"]
+        read_only_fields = ["id"]
+
+    def _get_or_create_ingredients(self, ingredients, recipe):
+        for ingredient in ingredients:
+            ingredient_obj, create = Ingredient.objects.get_or_create(**ingredient)
+            recipe.ingredients.add(ingredient_obj)
+
+    def create(self, validated_data):
+        ingredients = validated_data.pop("ingredients", [])
+        recipe = Recipe.objects.create(**validated_data)
+
+        self._get_or_create_ingredients(ingredients, recipe)
+
+        return recipe
+
+    def update(self, instance, validated_data):
+        ingredients = validated_data.pop('ingredients', None)
+        if ingredients is not None:
+            instance.ingredients.clear()
+            self._get_or_create_ingredients(ingredients, instance)
+
+        for attr, value in validated_data.items():
+            setattr(instance, attr, value)
+
+        instance.save()
+        return instance
```

Comment:
Although it's ok, the common practice here is to keep model "thin" and provide validaiton logic and creating models not in serailizer, but in service layer i.e. `ingredients_service.py` which would do call models' methods like `Ingredient.objects.create()` etc.

Reply:
Fair point, Krzys! I was talking with Victor about it and he said that this refactory is beyond of the scope of the exercise have in mind it will require a whole infrastructure to do it. Thanks for all technical guidelines that you sent me:)